text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,485 @@
+
+For example, this self type is valid:
+
+```rust
+impl MyType {
+     fn by_rc_to_box(self: Rc<Box<Self>>) { ... }
+}
+```
+
+The Rust language doesn't provide a way for user code to use this recursive property in generics or iteration, so this trait is unlikely to be useful except to the compiler. Nevertheless, we don't intend to _prevent_ use of the `Receiver` trait by user code: since the same recursive property applies to `Deref` yet it's been occasionally useful to [introduce `Deref` bounds](https://doc.rust-lang.org/std/pin/struct.Pin.html#method.new_unchecked).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `core` libs changes
+
+The `Receiver` trait is made public (removing its `#[doc(hidden)])` attribute), exposing it under `core::ops`. It gains a `Target` associated type.
+
+This trait marks types that can be used as receivers other than the `Self` type of an impl or trait definition.
+
+```rust
+pub trait Receiver {
+    type Target: ?Sized;
+}
+```
+
+A blanket implementation is provided for any type that implements `Deref`:
+
+```rust
+impl<P: ?Sized> Receiver for P
+where
+    P: Deref,
+{
+    type Target = <P as Deref>::Target;
+}
+```
+
+(See [alternatives](#no-blanket-implementation) for discussion of the tradeoffs here.)
+
+It is also implemented for `&T`, `&mut T`, `*const T` and `*mut T`.
+
+## Compiler changes
+
+The existing Rust [reference section for method calls describes the algorithm for assembling method call candidates](https://doc.rust-lang.org/reference/expressions/method-call-expr.html). This algorithm changes in one simple way: instead of dereferencing types (using the `Deref<Target=T>`) trait, we use the new `Receiver<Target=T>` trait to determine the next step.
+
+(Note that the existing algorithm isn't quite as simple as following the chain of `Deref<Target=T>`. In particular, `&T` and `&mut T` are considered as candidates too at each step; this RFC does not change that.)
+
+Because a blanket implementation is provided for users of the `Deref` trait and for `&T`/`&mut T`, the net behavior is similar. But this provides the opportunity for types which can't implement `Deref` to act as method receivers.
+
+## Object safety
+
+Receivers are object safe if they implement the (unstable) `core::ops::DispatchFromDyn` trait.
+
+As not all receivers might want to permit object safety or are unable to support it, object safety should remain being encoded in a different trait than the here proposed `Receiver` trait, likely `DispatchFromDyn`.
+
+This RFC does not propose any changes to `DispatchFromDyn`. Since `DispatchFromDyn` is unstable at the moment, object-safe receivers might be delayed until `DispatchFromDyn` is stabilized. `Receiver` is not blocked on further `DispatchFromDyn` work, since non-object-safe receivers already cover a big chunk of the use-cases.
```

Comment:
Rust for Linux also needs dyn upcasting, so we are using unstable `DispatchFromDyn` trait as well. But it sounds like that is out of scope for this RFC.

Reply:
Dyn upcasting should be stabilized soon. It was scheduled for next release, but [had to be reverted due to a soundness bug](https://github.com/rust-lang/rust/issues/120222).

Reply:
`DispatchFromDyn` and dyn upcasting are two different things. Dyn upcasting is `dyn TraitSub` => `dyn TraitSuper`. This was scheduled to be stabilized for the next release. `DispatchFromDyn` is what allows calling a method on a trait object. This one wasn't being stabilized, nor is it covered by this RFC. It's implementation is also rather fragile and not ready to allow arbitrary user implementations. (your type needs to have a very specific shape to not get an ICE when trying to use `DispatchFromDyn` on it)

Reply:
@bjorn3 Thanks for clarifying. The one we need is `DispatchFromDyn`, and we are careful to only use it for structs of a specific shape.